First draft of dashboard

LiveNathan opened the pull request:
There's lot of code in here I'm not using, yet. Some of it is from JS101 and I'm waiting to see if we'll use it. The first two functions I left in but commented out because you asked us to write them in an earlier lesson.

LiveNathan committed:
Add dotenv dependency and utilize .env for API keys

This commit adds the dotenv package to the project as a new dependency, allowing us to manage environment variables more securely. The API keys for Ipstack and OpenWeatherMap have been moved to a new .env file which is now properly ignored by git for security reasons.

Also, code in main.js was updated to handle these changes, including adding utility functions to convert an object to a string and vice versa. Plus, weather data retrieval was updated to include the sunrise, sunset, and solar noon times in addition to the previously fetched data.

These changes increase overall security by avoiding hard-coded sensitive data and improve the code structure for better maintainability.

LiveNathan committed:
Refactor data communication and DOM manipulation logic

Refactored the methods for getting data from IP Stack and Countries APIs to return the data, instead of injecting the data directly into the DOM. This change decentralizes the responsibility of accessing data and manipulating the DOM, keeping the functions cleaner and more focused on their main tasks. Additional function, main(), is added to manage the flow of data.

LiveNathan committed:
Refactor data fetching functions for modularity

This commit updates three data-fetching functions to return data instead of directly modifying the DOM. By returning the data fetched, the functions `getWeatherData()`, `getIssLocationData()`, and `getIssPeopleData()` become more modular and reusable. This also better separates the concerns of data fetching and data display in the code.

LiveNathan committed:
Replace sun multiplier function with new utility functions

Removed an unnecessary sun multiplier function that was based on solar events like sunrise, solar noon, and sunset times. This was done to simplify the code and improve efficiency. Replaced this function with two general-purpose utility functions - one to calculate distance between latitudes and longitudes and another to compare Gini coefficients (often used in socio-economic analysis). Also, modified the script call in the index.html file to improve script loading.

LiveNathan committed:
Add handleIpStackData function to display ipstackData

This commit introduces a function 'handleIpStackData' designed to render ipstack related data from the IP API on the UI. It also includes a function 'createHTMLElement' that generates HTML elements dynamically. Both these changes aim to simplify and streamline the dynamic generation of details on the UI and improve data representation. Removed manual document element updates with the fetched data. As a part of this, the code that fetches and displays the IP related data is refactored for better manageability and clarity.

LiveNathan committed:
Refactor main.js for better readability and function reusability

This commit includes substantial changes to 'main.js' to improve readability and enhance code reusability. The most significant change includes removing some hard-coded fetch parameters and isolating them into their dedicated function. This allows for the reusability of these functions elsewhere.  Alongside, the main function has been converted to an asynchronous function, handling promises better. Two keys were moved into respective functions resulting in improved code organization. This restructuring facilitates more manageable debugging, scalable features, and cleaner code organization.

LiveNathan committed:
Refactor IP and weather data fetching, add population and improvement in dashboard

The main changes involve refactoring how we fetch data from the IP Stack and the Weather API.
Instead of storing all data into variables and console logging, we now directly return the fetched data, which simplifies the code and improves readability.

Also, some dashboard improvements were made, displaying more, and more clear information to the user. Now, to the fetched IP information, we add population data, which gives additional insight into the country data.

Weather data is also being used more effectively, providing current weather info to appear on the dashboard. This includes sunrise, sunset, temperature, and pressure, among other useful pieces of information.

LiveNathan committed:
Refactor code and improve visual layout of dashboard

Issues with current code ambiguity and lack of meaningful visual distinction on the dashboard prompted this refactor.

Merging 'latitude' and 'longitude' into a single element 'latAndLon' gives a cleaner look and reduces repetitions. Added wind speed to the dashboard in the weather section to provide more comprehensive information. Moved variable declarations inside 'getSun' and 'getIssLocationData' closer to where they're needed, improving readability. Refactored functions 'getDistanceFromLatLonInKm', 'getIssLocationData' and 'getIssPeopleData' to directly return the result/values which simplifies the code.

Removed 'createHTMLElement' function as it was redundant - replaced with 'createElementAndAppend' wherever used. Added 'getIssData' function to display ISS information on the dashboard. Changed layout of 'root' div by adding CSS classes in 'index.html' and 'main.js' to enhance visual separation of different sections on the dashboard. Entire dashboard should now appear more organized and visually clear.

LiveNathan committed:
Update world time display and refactor code

Updated the getWorldTime function to return an object with multiple time properties. This change supports the addition of the day of the year and week number to the displayed time information. Also, code refactoring was performed for a cleaner structure, including using constants instead of let, and separating the time display functionality into a separate function "addTimeToDashboard". This provides a cleaner code architecture and improves application performance.

LiveNathan committed:
Add sun information to the dashboard

In this commit, the main dashboard function `addTimeToDashboard(time)` has been augmented to also include information about sunrise, sunset, solar noon, and the length of the day. This information is separately obtained and calculated in the new `getSun()` function, and afterwards passed to `addTimeToDashboard(time, sun)`.

This additional information is useful to provide users more details about sun timings and day length, hence enhancing the dashboard functionalities. Some codes on weather information which are repetitive with the provided sun information were also commented out in `getCountryData(countryCode)`. Lastly, the display styling of body in index.html was improved for better appearance.

iansedano reviewed (approved):
Not much to say really, it's looking good for this stage of the proceedings!

LiveNathan committed:
Refine UI and fix typos in class names

This commit updates the dashboard interface for improved readability and correctly aligns sections by fixing typos in CSS class names.

In the main.js file, changes are made to enhance the design of IP address and weather information by changing the text color to grey. For each piece of displayed information, a comma is added after the latitude for clarity. Also, all incorrect references to CSS classes ('flew-wrap' corrected to 'flex-wrap') are corrected.

In the index.html file, design enhancements are made by applying shadow and padding to the root div.

These changes ensure users can read the information quickly and intuitively while making the interface visually pleasant and more professional.
